### PR TITLE
Use a separate configmap for Ingress Controller values

### PR DESCRIPTION
**Pull request by rossf7**
Towards giantswarm/giantswarm#4795

Follow up to https://github.com/giantswarm/cluster-operator/pull/778. We can't have the IC replicas in the cluster configmap because if the cluster is scaled all apps will be updated. This moves it to a 2nd configmap called `ingress-controller-values`.

```
$ k get cm -n w4uyj
NAME                        DATA   AGE
ingress-controller-values   1      10s
w4uyj-cluster-values        1      54s
```

The desired unit test gets dropped. It was already removed from the `clusterapi` controller and these UTs quickly become large and brittle.

An e2e test for the contents of the cluster namespace would be more useful.

### CODE REVIEW

File: pkg/v22/resource/clusterconfigmap/current.go [hunk 1]
```diff
@@ -45,19 +47,25 @@ func (r *StateGetter) GetCurrentState(ctx context.Context, obj interface{}) ([]*
...
 	}
 
-	name := key.ClusterConfigMapName(clusterConfig)
+	var configMaps []*corev1.ConfigMap
+	{
+		r.logger.LogCtx(ctx, "level", "debug", "message", fmt.Sprintf("finding configMaps in namespace %#q", key.ClusterID(clusterConfig)))
 
-	r.logger.LogCtx(ctx, "level", "debug", "message", fmt.Sprintf("finding cluster configMap %#q in namespace %#q", name, clusterConfig.ID))
+		o := metav1.ListOptions{
+			LabelSelector: fmt.Sprintf("%s=%s", label.ManagedBy, project.Name()),
```
Comment (rossf7) [inline pkg/v22/resource/clusterconfigmap/current.go]: We now get all configmaps with the label. This excludes user configmaps created via our API or with kubectl.

File: pkg/v22/resource/clusterconfigmap/desired.go [hunk 2]
```diff
@@ -46,26 +45,55 @@ func (r *StateGetter) GetDesiredState(ctx context.Context, obj interface{}) ([]*
...
-		"ingressController": map[string]interface{}{
-			"replicas": ingressControllerReplicas,
+	configMapSpecs := []configMapSpec{
+		{
+			Name:      key.ClusterConfigMapName(clusterConfig),
+			Namespace: key.ClusterID(clusterConfig),
+			Values: map[string]interface{}{
+				"baseDomain":   key.TenantBaseDomain(clusterConfig),
+				"clusterDNSIP": clusterDNSIP,
+				"clusterID":    key.ClusterID(clusterConfig),
```
Comment (rossf7) [inline pkg/v22/resource/clusterconfigmap/desired.go]: These values are static and can be used by all apps including optional apps.

File: pkg/v22/resource/clusterconfigmap/desired.go [hunk 3]
```diff
@@ -46,26 +45,55 @@ func (r *StateGetter) GetDesiredState(ctx context.Context, obj interface{}) ([]*
...
+		},
+		{
+			Name:      key.IngressControllerConfigMapName,
+			Namespace: key.ClusterID(clusterConfig),
+			Values: map[string]interface{}{
+				"baseDomain": key.TenantBaseDomain(clusterConfig),
+				"clusterID":  key.ClusterID(clusterConfig),
+				"ingressController": map[string]interface{}{
+					"replicas": ingressControllerReplicas,
+				},
```
Comment (rossf7) [inline pkg/v22/resource/clusterconfigmap/desired.go]: We need all these values not just the replicas. See https://github.com/giantswarm/nginx-ingress-controller-app/blob/0264ae24f53f5530bd07511d94ec95531d732a35/helm/nginx-ingress-controller-app/values.yaml#L9L21.

**Comment by puja108** [comment]
only one Q: so when there's 2 configmaps how does the merging look like?

**Comment by fiunchinho** [comment]
It looks good to me, but I'm not comfortable enough with my current knowledge to actually approve it.

**Comment by rossf7** [comment]
> so when there's 2 configmaps how does the merging look like?

We can only specify 1 app level configmap per CR. In the next PR I'll extend the `app` resource so the configmap can be set. If its unset we default to `eggs2-cluster-values` like now.

**Comment by kopiczko** [comment]
> > so when there's 2 configmaps how does the merging look like?
> 
> We can only specify 1 app level configmap per CR. In the next PR I'll extend the `app` resource so the configmap can be set. If its unset we default to `eggs2-cluster-values` like now.

This feel rather hacky. Won't we have more cases like that? I.e. something restarting because irrelevant values change and we need to extract that to a separate CM? Or I am making up a problem?

**Comment by rossf7** [comment]
We need to make sure the values in the cluster-values configmap are static for the lifetime of the cluster. Now it has been split they are.

It is a problem but for complex apps we may need a configmap for dynamic values. For Ingress Controller once we have confidence in running HPA for all clusters we will remove this.

@fiunchinho Also needs to add a configmap for cluster-autoscaler on Azure. This needs the max and min workers which is dynamic. Again this is temporary it should auto discover like on AWS. This is already merged upstream but isn't in a release yet.

**Pull request merged**